include/envoy/upstream/cluster_manager.h

```diff
@@ -147,6 +167,13 @@ class ClusterManager {
    * @return std::string the local cluster name, or "" if no local cluster was configured.
    */
   virtual const std::string& localClusterName() const PURE;
+
+  /**
+   * These two methods allow to register and un register callback for cluster lifecycle events
```

Comment:
nit: "unregister" and "callbacks"

Reply:
Also, can you flesh out the docs here about threading? I see from the code that this effectively needs to be called on every TLS thread to be useful. If we keep this method, I would probably rename to `addThreadLocalClusterUpdateCallbacks` etc. Though this is making me wonder if this is the most useful programming model. I wonder if it would be better for the callbacks to be registered on the main thread once, and then fired on all TLS workers? I suspect that might be easier to use in the general case.

Reply:
If you're a filter that is dynamically registering for the duration of a request, you only want the thread-local callback.

Reply:
I wasn't suggesting not having a thread local callback, but having to register for it on each worker seems odd. In general, a filter will register for callbacks on the main thread, and then expect to receive them on each worker. That's what I'm suggesting.

Reply:
The flow for this filter is that a request comes in for cluster X. Filter registers to hear about clusters while waiting for X to arrive and the unregisters after (it no longer cares about cluster updates). Are you suggesting that it would be preferable to always be listening for cluster updates?

Reply:
I see. There are use cases in which I think we do want to always listen and we will want to register on the main thread for updates to fire on each worker. We can start with this, but I would still clarify in the docs.

Reply:
I updated the docs.